### PR TITLE
Optional Dependencies

### PR DESCRIPTION
**Pull request by swainn**
This PR makes most Python dependencies optional.

TODO:

- [x] Tests and Coverage
- [x] Linting and Black
- [x] Documentation include instructions for dependency installation
- [x] Test manual production deployment
- [x] Update `tethys gen metayaml` to generate minimal and full versions of the environment
- [x] Update CI to build `micro-tethys-platform` package in addition to the "normal/full" build

### CODE REVIEW

**Comment by swainn** [comment]
Add instructions to the documentation for installing dependencies for the following features:

- [ ]

File: docs/tethys_sdk/gizmos/bokeh_view.rst [hunk 1]
```diff
@@ -2,7 +2,24 @@
...
+
+   .. code-block:: bash
+
+      # conda: conda-forge channel strongly recommended
+      conda install -c conda-forge "bokeh<3"
+
+      # pip
+      pip install "bokeh<3"
+
+   **Don't Forget**: If you end up using this gizmo in your app, add `bokeh` as a requirement to your `install.yml`.
```
Comment (swainn) [inline docs/tethys_sdk/gizmos/bokeh_view.rst:19]: @sdc50 I took a first pass at a dependency install tip box in docs for optional dependencies. What do you think of this?

Comment (swainn) [reply docs/tethys_sdk/gizmos/bokeh_view.rst:19]: Looks like this:

image

Comment (sdc50) [reply docs/tethys_sdk/gizmos/bokeh_view.rst:19]: This looks great! 

As a side note I think we can take off the pin for Bokeh. I've tested it with v3 and it still works great.

Comment (sdc50) [reply docs/tethys_sdk/gizmos/bokeh_view.rst:19]: One other thought, do we want to mention anything about Tethys v5.0?

**Comment by swainn** [comment]
Tutorial language: "...this package may not be included in some Tethys installations, so you should include it as a dependency of your app to ensure it is installed."

File: docs/tethys_sdk/gizmos/bokeh_view.rst (hunk 1 above)
Comment (swainn) [reply docs/tethys_sdk/gizmos/bokeh_view.rst:19]: "This gizmo requires the `bokeh` library to be installed. Starting with Tethys 5.0 or if you are using `microtethys`, you will need to install `bokeh` using conda or pip as follows:"

Comment (swainn) [reply docs/tethys_sdk/gizmos/bokeh_view.rst:19]: > As a side note I think we can take off the pin for Bokeh. I've tested it with v3 and it still works great.

I get an error in the showcase app when I have Bokeh 3 installed. See: #980

**Comment by sdc50** [comment]
@swainn We need to rebase this branch on master. I'm inclined to squash the commits and then rebase and force push. That would require you (and anyone else that has this branch checked out) to re-set it (or delete and re-checkout). Do you have any changes to push before I do that?

**Comment by swainn** [comment]
> @swainn We need to rebase this branch on master. I'm inclined to squash the commits and then rebase and force push. That would require you (and anyone else that has this branch checked out) to re-set it (or delete and re-checkout). Do you have any changes to push before I do that?

Sounds like a good plan to me. I don't have any changes that I need to merge.

File: docs/supplementary/optional_features.rst [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+With the release of ``microtethys`` and starting with Tethys v5.0 many features of Tethys Platform will become optional and require additional dependencies to be installed. This is done to limit the size of the environment and allow Tethys Portal administrators more flexibility in deciding what features are needed in their deployment. The following list of features are optional and will require the listed dependencies to be installed for that feature to be enabled:
+
+Security Features
+=================
+
+Session Security
+----------------
+
+Session security enables setting timeouts for user sessions.
```
Comment (swainn) [inline docs/supplementary/optional_features.rst:15]:
```suggestion
Session security enables setting timeouts for user sessions and automatically logging them out.
```

File: docs/supplementary/optional_features.rst [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+**dependencies**
+ - ``django-session-security``
+
+
+Track Failed Login Attempts
+---------------------------
+
+Tracking failed loging allows Tethys to lock user accounts after a certain number of failed attempts, and alerts users of the number of failed attempts when they login.
```
Comment (swainn) [inline docs/supplementary/optional_features.rst:24]:
```suggestion
Tracking failed logins allows Tethys to lock user accounts after a certain number of failed attempts, and alerts users of the number of failed attempts when they login.
```

File: docs/supplementary/optional_features.rst [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+**dependencies**
+    - ``django-oauth-toolkit``
+
+Portal Enhancements
+===================
+
+Terms and Conditions
+--------------------
+
+Enables portal administrators to define terms and conditions that users must accept to user the portal.
```
Comment (swainn) [inline docs/supplementary/optional_features.rst:113]:
```suggestion
Enables portal administrators to define terms and conditions that users must accept to use the portal.
```

File: docs/supplementary/optional_features.rst [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+**dependencies**
+    - ``condorpy``
+
+External Services
+=================
+
+Dataset Services
+----------------
+
+Enables the :term:`dataset services` API.
```
Comment (swainn) [inline docs/supplementary/optional_features.rst:248]:
```suggestion
Enables the :term:`dataset services` APIs for CKAN and GeoServer.
```

**Review by swainn — COMMENTED** [review]
Aside from my other comments I have one other concern: how are we testing the micro version / optional dependencies. I think we decided that it was too complex to test both ways, but I wonder if there is a straightforward way to modify tests that are affected by an optional dependency. Something like a `try: except:` that catches the exception raised? Maybe as a decorator so we can easily add it to the tests that need it. Have you tried anything like that?

**Comment by sdc50** [comment]
fixes #980 and fixes #985

**Comment by sdc50** [comment]
@swainn I've finished the docs for this PR, so it should be ready to go.

**Review by swainn — COMMENTED** [review]
I Approve

**Pull request merged**